lib/Dialect/TritonGPU/Transforms/UpdateMmaForVolta.cpp

```diff
@@ -0,0 +1,349 @@
+#include "Utility.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "triton/Analysis/Utility.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h"
+
+namespace mlir {
+namespace {
+using triton::DotOp;
+using triton::gpu::ConvertLayoutOp;
+using triton::gpu::DotOperandEncodingAttr;
+using triton::gpu::MmaEncodingAttr;
+using triton::gpu::SliceEncodingAttr;
+
+// This pattern collects the wrong Mma those need to update and create the right
+// ones for each.
+class CollectMmaToUpdateForVolta : public mlir::RewritePattern {
+  // Holds the mapping from old(wrong) mmaEncodingAttr to the new(correct)
+  // mmaEncodingAttr.
+  DenseMap<MmaEncodingAttr, MmaEncodingAttr> &mmaToUpdate;
+
+public:
+  CollectMmaToUpdateForVolta(
+      mlir::MLIRContext *ctx,
+      DenseMap<MmaEncodingAttr, MmaEncodingAttr> &mmaToUpdate)
+      : mlir::RewritePattern(triton::DotOp::getOperationName(), 1, ctx),
+        mmaToUpdate(mmaToUpdate) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::Operation *op,
```

Comment:
If it always returns failure, does it mean `matchAndRewrite` is only called once? If so, maybe we don't need to declare `CollectMmaToUpdateForVolta` as a subclass of `RewritePattern`?

Reply:
> If it always returns failure, does it mean `matchAndRewrite` is only called once? If so, maybe we don't need to declare `CollectMmaToUpdateForVolta` as a subclass of `RewritePattern`?

Yeah, maybe a method of UpdateVoltaPass is better.

Reply:
I tried to replace this Pattern with a `module.walk()` in [https://github.com/openai/triton/pull/1048/commits/f161fcbc52f81b8314f70502f84eaf94c964685b](commit), but get an error in `applyPatternsAndFoldGreedily` on assert-enabled-LLVM, seems that the mixing execution of `walk` and `RewritePattern` cause some issues.
Couldn't get it happy with a quick fix, so I left a TODO here to solve it later.

Reply:
Could you use [applyOpPatternsAndFold](https://mlir.llvm.org/doxygen/namespacemlir.html#aa858fc38fd77156306692e198959ecd4) though? Not as good as `walk`, but doesn't do anything recursively so it would be marginally cleaner

Reply:
It seems that the `applyOpPatternsAndFold` is not what we want, referring to its doc
> Applies the specified patterns on op alone while also trying to fold it
It simply applies on a single op passed in, we still need a driver to make it traverse on a region.

We might still need the `applyPatternsAndFoldGreedily` although it doesn't actually recursively called, it helps to traverse the ops in regions.